### ISSUE TITLE
`Should-Throw`: How to negate the evaluation?

### ISSUE THREAD

**Issue by johlju** [pester/Pester#2528]
### Checklist

- [X] Issue has a meaningful title
- [X] I have searched the existing issues. See [all issues](https://github.com/pester/Pester/issues?q=is%3Aissue)
- [X] I have tested using the latest version of Pester. See [Installation and update guide](https://pester.dev/docs/introduction/installation).

### What is the issue?

In Pester 5 syntax we could say `Should -Not -Throw`. I can't find a way to negate the `Should-Throw` command as it does not have a `Not` parameter or a command `Should-NotThrow`.

### Expected Behavior

Possibility to evaluate a command to _not_ throw.

### Steps To Reproduce

not applicable

### Describe your environment

```
Pester version     : 6.0.0-alpha3 /Users/johlju/source/PesterConverter/output/RequiredModules/Pester/6.0.0/Pester.psm1  
PowerShell version : 7.4.3
OS version         : Unix 14.5.0
```

### Possible Solution?

Either add a `Not` parameter to the command `Should-Throw` or add a new command `Should-NotThrow`?

**Comment by fflaten** [comment]
`-Not -Throw` is currently intentionally left out. We kinda discussed this on discord in May when you requested more verbose error output. 😃 

> How about running the code without `Should -Not -Throw`? The exception itself will fail the test and show full stack. Not throw mostly provides a simple error message which you found too simple.

Did you find a scenario where `Should -Not -Throw` still makes sense?

**Comment by johlju** [comment]
> We kinda discussed this on discord in May 

Right... remember that. 🙂  Didn't thought it would apply here. 😉 

> Did you find a scenario where Should -Not -Throw still makes sense?

Only when automating conversion of test scripts. E.g. only in SqlServerDsc there are 788 lines using the syntax 
```{ ... } | Should -Not -Throw```. Changing them to not pass a scriptblock through the pipeline and not use `-Not` manually would take a long time. I have now code ready that can easily convert Pester 5 syntax to either `Should-NotThrow` or `Should-Throw` with `-Not`. If we are not considering adding either the command `Should-NotThrow` or have `-Not` added to `Should-Throw` then I need to look into code that can convert all ```{ ... } | Should -Not -Throw``` by just running the code inside the scriptblock and without passing it to `Should`. 🤔

**Comment by fflaten** [comment]
AFAIK it's not set in stone, but by starting without we at least get the discussion. If it can be covered by a simple migration step trough docs, made even easier by your automated module, that'd be cool.

The easiest conversion is probably to replace `Should -Not -Throw` with `ForEach-Object { & $_ } | Out-Null` as it would also support the undocumented support for multiple scriptblocks through pipeline. E.g.

```powershell
# Defining outside to be available for test in the end
$scriptBlock = { throw 'one' }

$source = {
    $scriptBlock | Should -Not -Throw

    { throw 'two' } | Should -Not -Throw -Because 'why not'

    { throw 'three' } |
    Should -Not -Throw -Because 'why not' -ExpectedMessage 'ignore'

    Should -Not -Throw -Because 'why not' -ExpectedMessage 'ignore' -ActualValue { throw 'four' }

    # I'm sure there's one out there :D
    "throw 'five'" | ForEach-Object { [scriptblock]::Create($_) } | should -Throw -Not
}

$Ast = $source.Ast
$todo = $Ast.FindAll({ param($a) $a -is [System.Management.Automation.Language.CommandAst] -and $a.GetCommandName() -eq 'Should' }, $true)
$todo | ForEach-Object {
    $pipelineElements = $_.Parent.PipelineElements
    $inputScriptBlock = $null
    if ($pipelineElements.Count -eq 1) {
        # Should -Not -Throw -ActualValue { something }
        # AFAIK positional doesn't work so being naive. 99% probably use pipeline input like docs
        $inputScriptBlock = for ($i = 0; $i -lt $_.CommandElements.Count; $i++) {
            $p = $_.CommandElements[$i]
            if ($p -is [System.Management.Automation.Language.CommandParameterAst] -and $p.ParameterName -like 'A*') {
                $_.CommandElements[$i + 1]
                break
            }
        }
    } elseif ($pipelineElements.Count -gt 1 -and $pipelineElements[-1] -eq $_) {
        # something | maybe something else | Should -Not -Throw
        $replaceStart = $_.Extent.StartOffset - $_.Parent.Extent.StartOffset
        $inputScriptBlock = $_.Parent.Extent.Text.Remove($replaceStart) -replace '\s*\|\s*$'
    } else {
        throw 'What sorcery is this?'
    }

    # New parent extent
    $newParentExtentText = "$($inputScriptBlock) | ForEach-Object { & `$_ } | Out-Null # Converted from Should -Not -Throw"
    # Test
    [System.Management.Automation.Language.Parser]::ParseInput($newParentExtentText, [ref]$null, [ref]$null).GetScriptBlock().Invoke()
}
```

**Comment by johlju** [comment]
Thanks for the code example, and the suggestions for use of `ForEach-Object`, did not think of that. It is the safest (only) option if scriptblocks are passed as variables to `Should`. Maybe I will try to have both options 1) using `ForEach-Object` when there are a variable and 2) if a scriptblock, parse out the code in the scriptblock so it can run directly in the It-block as it is currently meant to do in Pester 6 alpha. 🤔 I will see how it works out. 🙂 

Btw. using `ForEach-Object` outputs a bit more line numbers (first one) then having the code directly in the scriptblock (second one). So to avoid using `ForEach-Object` would be cleaner I think. 🤔 
 
```powershell
It 'Should not throw, passing to ForEach-Object' {
    {
        Write-Error -Message 'MockErrorMessage' -ErrorId 'MockErrorId' -Category 'InvalidOperation' -TargetObject 'MockTargetObject' -ErrorAction 'Stop'
    } | ForEach-Object -Process { & $_ } | Out-Null
}

It 'Should not throw, without any scriptblock' {
    Write-Error -Message 'MockErrorMessage' -ErrorId 'MockErrorId' -Category 'InvalidOperation' -TargetObject 'MockTargetObject' -ErrorAction 'Stop'
}
```

Outputs:

```pester
[-] Should not throw, passing to ForEach-Object 19ms (14ms|5ms)
 WriteErrorException: MockErrorMessage
 at <ScriptBlock>, /Users/johlju/source/PesterConverter/tests/Unit/Pester6.tests.ps1:42
 at <ScriptBlock>, /Users/johlju/source/PesterConverter/tests/Unit/Pester6.tests.ps1:43
 at <ScriptBlock>, /Users/johlju/source/PesterConverter/tests/Unit/Pester6.tests.ps1:39
[-] Should not throw, without any scriptblock 10ms (6ms|4ms)
 WriteErrorException: MockErrorMessage
 at <ScriptBlock>, /Users/johlju/source/PesterConverter/tests/Unit/Pester6.tests.ps1:57
```

**Comment by fflaten** [comment]
You can invoke a scriptblock-variable with call operator. E.g. `$null = & (<actualvalue>)` should work for variables, script blocks and expressions (note parentheses)
 
Only reason I didn't use it was the edge case of multiple script blocks, which you can probably ignore.

**Comment by nohwnd** [comment]
Thanks for raising the issue, and for working out a workaround. Closing because currently the plan is to not re-introduce Should-NotThrow.

**Comment by fflaten** [comment]
@nohwnd What about soft assertions?

**Comment by nohwnd** [comment]
> [@nohwnd](https://github.com/nohwnd) What about soft assertions?

Does this comment belong here?

**Comment by fflaten** [comment]
Yes, sorry for being unclear. Without a `Should-NotThrow`, how would we support soft assertion?

```powershell
$c = New-PesterConfiguration
$c.Should.ErrorAction = 'Continue'
$c.Run.ScriptBlock = {
    Describe 'd' {
        It 'i' {
            # { throw '123' } | Should -Not -Throw  # soft assertion works
            & { throw '123' } # soft assertion doesn't work
            1 | Should-Be 2
            2 | Should-Be 3
            "right" | Should-Be "wrong"
        }
    }
}

Invoke-Pester -Configuration $c
```

**Comment by nohwnd** [comment]
I see. I don't think soft assertions bring a lot of value here. Should -Not -Throw is implied by every line of code that runs. If the code throws in unexpected way it should take the test down, and I have hard time imagining situation where getting more info from the test helps debugging the problem. 

If the code was supposed to not throw and it threw exception. It makes sense that all subsequent assertions are probably testing the wrong thing and will fail as well. 

I would like a very compelling example before re-introducing Should-NotThrow, because it leads to lazy tests that check that code passes, but not that it is functional (e.g. returns the correct values).

**Comment by fflaten** [comment]
I agree in general. Wouldn't be surprised if it is used in integration tests though. 

Like a deployment or installation failed unexpectedly, but did it at least cleanup as expected (function has a `try/finally`? No leftovers? Environment variables restored? Created the log file, so we can capture it etc. Something like https://github.com/PowerShell/PowerShell/blob/master/test/packaging/windows/msi.tests.ps1#L378-L394

Most use cases can be resolved with a different test design/structure, or using `try/finally`. I'm fine with documenting any migration tips for now in pester/docs#411, like writing minimal tests and cleanup in `try/finally` or `AfterEach`. Just wanted to make sure we'd considered it 🙂

**Comment by nohwnd** [comment]
I would not be surprised if it is used either, but I still don't see big value of it. :)

What you describe reads (to me) like a different test that is worth writing explicitly if we care about that. And the test would use Should-Throw, and not Should-NotThrow. Because we expect a failure and then we expect the result being that everything is clean. 

If we do expect the same result in both cases, then we should still validate that the code sometimes fails and sometimes does not fail. 

my action that did not fail
assert stuff was cleaned up

and

{ my action that fails } | should-throw
assert stuff was cleaned up

In the positive case, the my `action that did not fail` is not special, and when it fails it should tear down the test. That line randomly failing is not a replacement for the other test where it always fails (or provingly fails 5 out of 10 times etc.).
